---
title: "Improved Off-policy Reinforcement Learning in Biological Sequence Design"
date: 2026-02-16
permalink: /posts/2026/02/RL_seq/
tags:
  - deep-learning
  - Reinforcement-Learning
  - Biological-Sequence-Design
  - Proxy-training
  - ùúπ-Conservative-search
  - GFlowNets
math: true
---

## Abstract

Designing biological sequences such as DNA, RNA, and proteins is a central challenge in biotechnology due to the Huge search space and the high cost of estimation. Recent reinforcement learning (RL) approaches rely on proxy models to estimate quality, but these models often fail when exploring regions with unknown data. This paper introduces Œ¥-Conservative Search (Œ¥-CS), a novel off-policy reinforcement learning strategy that balances exploration and reliability in biological sequence design. The method focus on high-scoring known sequences by randomly masking tokens and then denoising them using a generative policy, ensuring that new candidates remain close to reliable regions of the search space. The degree of conservativeness is adaptively adjusted based on the proxy model‚Äôs uncertainty. Experiments across DNA, RNA, and protein design tasks demonstrate that Œ¥-CS consistently discovers higher quality sequences than existing methods. This makes Œ¥-CS a scalable and reliable framework for real-world biological design.

## Table of Contents

- [Introduction and Motivation](#introduction-and-motivation)  
- [Reinforcement Learning for Sequence Design](#reinforcement-learning-for-sequence-design)
- [The Core Problem: Proxy Misspecification](#the-core-problem-proxy-misspecification)  
- [Key Idea: Delta Conservative Search](#key-idea-delta-conservative-search)  
- [Adaptive Conservativeness Using Uncertainty](#adaptive-conservativeness-using-uncertainty)  
- [Experimental Setup and Benchmarks](#experimental-setup-and-benchmarks)  
- [Results](#results)  
- [Conclusion](#conclusion)

## Introduction and Motivation

In Biotechnology, designing biological sequences is a central problem.‚Ä®For example, we may want to design DNA sequences with high binding affinity, RNA molecules with specific folding properties, or proteins with enhanced fluorescence or stability. However, this is extremely challenging. The search space for the sequences grows exponentially Large with increasing sequence length, which makes brute-force search impossible.‚Ä®Furthermore, evaluating these sequences using an evaluating function, usually known as "Oracle function" or simply what we call "wet-lab experiments" is very expensive and takes a lot of time. Also, evaluation using Detailed simulation is not very cost and time effective as well.

To address this problem, we often use machine learning models called proxy models. But these proxies are imperfect, they are not always accurate and can make unreliable predictions, especially for unknown sequences.‚Ä®This paper addresses exactly this issue by proposing a conservative search strategy that improves robustness without sacrificing exploration. It introduces a novel off-policy search method called Delta-Conservative Search (Œ¥-CS), which is designed to enhance the robustness of generative policies, specifically Generative Flow Networks (GFlowNets), for designing biological sequences such as DNA, RNA, and proteins<a href="#ref-1" title="(2025) Improved Off-policy Reinforcement Learning in Biological Sequence Design">[1]</a>.

## Reinforcement Learning for Sequence Design

Biological sequence design can be framed as a sequential decision-making problem, we build a sequence one token at a time, and at the end we receive a score that reflects how good that sequence is according to a set objective or goal. This process makes reinforcement learning (RL) a perfect fit.

In this setting, an agent generates a sequence step by step. Each action corresponds to choosing the next token, for example; a nucleotide or amino acid. And once the full sequence is constructed, it is evaluated by an oracle (often an expensive wet-lab experiment or a high-fidelity simulator). Because querying the oracle is costly, modern approaches rely on proxy models that approximate the oracle and provide fast reward estimates during training itself.

### On-policy Reinforcement Learning

Early RL-based approaches for sequence design relied on on-policy methods, meaning the policy is trained only on data generated by its current version. A representative example is DyNA PPO, which applies Proximal Policy Optimization (PPO) within an active learning loop. On-policy methods have some appealing properties. They are relatively stable, simple, and align well with classical RL theory. However, they come with a major drawback which is poor data efficiency. Since they cannot effectively reuse previously collected sequences, large amounts of expensive oracle evaluations are needed to make progress. This drawback becomes highly problematic when dealing with Huge search spaces, where exploring is essential.

### Off-policy Reinforcement Learning

To address the limitations of On-policy RL, researchers have turned to off-policy RL methods, which can learn from data generated by older policies or even from fixed offline datasets. This is very effective in biology, where historical datasets and previously evaluated sequences are often available.

One effective off-policy framework for sequence design is Generative Flow Networks (GFlowNets). Instead of learning a single optimal sequence, GFlowNets learn a distribution over sequences, where the probability of sampling a sequence is proportional to its reward. This allows them to generate high-quality sequences , a crucial property when exploring biological landscapes. Another advantage of GFlowNets is their flexibility, they can mix offline data (existing evaluated sequences) with newly generated samples, leading to more stable and efficient training compared to on-policy RL methods.


## The Core Problem: Proxy Misspecification

Despite their promising performance, off-policy methods introduce a serious challenge, which is proxy model misspecification. Proxy models are trained on limited data and tend to make unreliable predictions on out-of-distribution (OOD) sequences. Off-policy methods, especially those that explore the search space widely, are more likely to generate such sequences.

This can lead to assigning greater reward to the sequence which may look promising but perform poor during evaluation by oracle function. The policy learns to exploit weaknesses in the proxy model rather than discovering meaningful biological sequences. Recent studies have shown that this issue becomes especially severe in large scale problems such as protein design, where the proxy model is particularly uncertain early in training. This creates a clear trade-off between exploration and reliability, which existing methods fail to balance effectively.

## Key Idea: Delta Conservative Search

The central idea of this paper is Delta-Conservative Search, abbreviated as Œ¥-CS. It is designed specifically to improve off-policy reinforcement learning methods like GFlowNets when proxy models are unreliable.‚Ä®The key intuition is simple, instead of exploring the entire sequence space blindly, we restrict exploration to regions where the proxy is more trustworthy. 

Œ¥-CS operates by starting from high-scoring sequences in the offline dataset and injecting noise in those sequences. Each token in a sequence is independently masked with probability Œ¥, where Œ¥ controls the degree of conservativeness. A smaller Œ¥ results in minor changes in the original sequence, while a larger Œ¥ allows more exploration as it changes the sequence to a larger extent. The masked sequences are then denoised sequentially using a GFlowNet policy, producing new candidate sequences that remain close to known high-quality sequences.

To further improve effectiveness of the search, Œ¥-CS adapts Œ¥ based on the uncertainty of the proxy model. When the proxy is uncertain, exploration is automatically restricted and when it is confident, broader exploration is allowed. This adaptive conservativeness enables Œ¥-CS to balance novelty and reliability, leading to more stable training and higher-quality sequence discovery across diverse biological design tasks.

### 1. Problem Formulation

We want to find sequences x ‚àà V<sup>L</sup>, Where V is the vocabulary (eg. amino acids or nucleotides) and L is the sequence length. We also look for some desired properties from these sequences such as binding affinity or enzymatic activity etc. These properties are evaluated by a black box oracle function ; f : V<sup>L</sup> ‚üπ R <a href="#ref-2" title="(2022) Biological sequence design with GFlowNets. In International Conference on Machine Learning (ICML)">[2]</a><a href="#ref-8" title="(2020) Model-based reinforcement learning for biological sequence design. In International Conference on Learning Representations (ICLR)">[8]</a>. Evaluation of this function (f) is very costly and time consuming because it involves wet-lab experiments and simulations. 

### 2. Active Learning for Biological Sequence Design

Biological sequence design often face problems due to limited experimental budgets, which makes it difficult to evaluate large numbers of candidate sequences using expensive laboratory testings. To address this, this paper adopts an active learning framework that iteratively improves both the predictive model and the generative policy using a small number of carefully selected queries. We will use some previous studeies for the Active Learning<a href="#ref-2" title="(2022) Biological sequence design with GFlowNets. In International Conference on Machine Learning (ICML)">[2]</a>.

![Figure 1: Active Learning for biological sequence design.]({{ site.baseurl }}/images/Fig1.png)
*Figure 1: Active Learning for biological sequence design.*


The active learning process has multiple rounds. In each round, a proxy model is trained on the currently available dataset (<i>Fig 1 : Step A</i>) to approximate the true oracle that evaluates sequence quality. This proxy enables rapid reward estimation and guides the search process. Next, a generative policy (<i>Fig 1 : Step B</i>), implemented using Generative Flow Networks (GFlowNets), is trained to propose promising new sequences based on the proxy‚Äôs predictions. Most important, this step intergates Œ¥-Conservative Search with GFlowNets, ensuring that proposed sequences remain within trustworthy regions of the search space.

Finally, a batch of generated sequences is evaluated using the true oracle, and the resulting data is added to the dataset for the next round (<i>Fig 1 : Step C</i>). By repeating this cycle, the framework efficiently balances exploration and exploitation, discovering higher-quality biological sequences while minimizing costly oracle evaluations.

### 3. Proxy Training

We train a proxy model fœï(x) using the offline dataset D<sub>t‚àí1</sub> at round t. We focus on minimising the mean squared error loss. We follow previos studies<a href="#ref-2" title="(2022) Biological sequence design with GFlowNets. In International Conference on Machine Learning (ICML)">[2]</a> to use the initial dataset D<sub>0</sub>.

$$
\mathcal{L}(\phi)=\mathbb{E}_{x \sim P_{D_{t-1}}(x)}\left[\left( f(x) - f_{\phi}(x) \right)^2\right]
$$

### 4. Policy Training with Œ¥-CS

The policy responsible for generating new biological sequences is trained using Generative Flow Networks (GFlowNets), an off-policy reinforcement learning framework designed to sample diverse high quality sequences. In this setting, the policy learns to generate sequences with probability proportional to a reward which is provided by proxy model.

$$
p(x;\theta) \propto R(x;\phi) = F(x;\phi)
$$

GFlowNets are generative models that learn a probability distribution over solutions rather than optimizing for a single best sequence. Each partial sequence is treated as a state, and actions correspond to adding tokens.‚Ä®Once a complete sequence is formed, it receives a reward. But, a major challenge in policy training is that proxy models can be highly unreliable for out-of-distribution (OOD) sequences, which can destabilize learning. To solve this problem, Œ¥-Conservative Search (Œ¥-CS) is integrated directly into policy training. Instead of training only on fully generated sequences, the policy is trained on trajectories obtained by first disturbing high-scoring offline sequences and then reconstructing them.

### 5. Policy and Learning Objective.

The policy consists of two components, a forward policy (P<sub>F</sub>) and a backward policy (P<sub>B</sub>). The forward Policy generates sequences token by token. It starts with an empty sequence (s<sub>0</sub>). After L steps we get the full sequence (s<sub>L</sub> = x), where L is the sequence length. The backward policy is used for modelling the probability of backtracking from the terminal state. The backward policy is deterministic, which simplifies training and makes GFlowNets equivalent to certain soft off-policy RL methods.

$$
P_F(\tau; \theta) = \prod_{i=1}^{L} P_F(s_i \mid s_{i-1}; \theta)
$$

The policy is trained to minimise the Trajectory balance (TB) Loss, which enforces consistency between forward and backward probabilities and rewards. However, the main challenge is that proxy models can produce highly unreliable rewards for out-of-distribution sequences.‚Ä®Training on such trajectories can destabilize learning and degrade performance <a href="#ref-2" title="(2022) Biological sequence design with GFlowNets. In International Conference on Machine Learning (ICML)">[2]</a>. The Œ¥-CS comes into play here.

$$
\mathcal{L}_{\mathrm{TB}}(\tau; \theta) = \left(\log \frac{Z_\theta \, P_F(\tau; \theta)}{R(x; \phi)}\right)^2
$$

### 6. Œ¥-Conservative Search (Œ¥-CS)

Œ¥-CS is an off policy search method which modifies off-policy trajectory generation and enables controllable exploaration in the search space through a control parameter Œ¥. The Œ¥ defines the probability of masking the tokens in the noise injection step. The Algorithm proceeds as follows :-

#### 6.1 Sampling high scoring offline sequences

To start from reliable points, sequences are sampled from a rank-based reweighted prior. Ee sample a reference sequence x from the prior distribution P<sub>D<sub>t-1</sub></sub>. Higher-ranked sequences are always assigned greater weights. This biases exploration toward promising regions while still allowing diversity. We use rank-based prioritization <a href="#ref-3" title="(2020) Sample-efficient optimization in the latent space of deep generative models via weighted retraining. In Advances in Neural Information Processing Systems (NeurIPS)">[3]</a>

$$
w(x; \mathcal{D}_{t-1}, k) \propto \frac{1}{kN + \mathrm{rank}_{f,\mathcal{D}_{t-1}}(x)}
$$

#### 6.2 Injecting Noise

In this step we introduce some noise into the obtained sequence from previosu step. In this step we use a noise injection policy which masks each individual token with a probability Œ¥. This ultimately determines how much of the original sequence is preserved. Smaller Œ¥ values result in conservation , while larger values encourage more exploration (Œ¥ ‚àà [0,1]). 

$$
P_{\text{noise}}(\tilde{x} \mid x, \delta) = \prod_{i=1}^{L} \left[\delta \cdot \mathbb{I} \{\tilde{e}_i = \text{[MASK]} \}+(1 - \delta) \cdot \mathbb{I}
\{\tilde{e}_i= e_i \}\right]
$$

#### 6.3 Denoising

In this step, the masked sequences from the previous step is reconstructed sequentially using the GFlowNet forward policy by predicting tokens from left to right.
The probability of denoising next token ƒï<sub>t</sub> from previously denoised subsquence ≈ù<sub>t‚àí1</sub> is :

$$
P_{\text{denoise}}(\hat{x} \mid \tilde{x}; \theta) =
\prod_{t=1}^{L}
\begin{cases}
\mathbb{I}\{\hat{e}_t = \tilde{e}_t\}, & \tilde{e}_t \neq \text{[MASK]}, \\[6pt]
P_F(\hat{s}_t \mid \hat{s}_{t-1}; \theta), & \tilde{e}_t = \text{[MASK]}.
\end{cases}
$$


After that, to obtain a fully reconstructed sequence $\hat{x}$ = ≈ù<sub>L</sub> we sample from denoising policy. The GFlowNet policy reconstructs masked tokens one by one from left to right, producing new sequences $\hat{x}$ whose level of novelty is controlled by the conservativeness parameter Œ¥.

$$
P_{\text{denoise}}(\hat{x} \mid \tilde{x}; \theta) = \prod_{t=1}^{L} P_{\text{denoise}}(\hat{e}_t \mid \hat{s}_{t-1}, \tilde{x}; \theta).
$$

## Adaptive Conservativeness Using Uncertainty

Choosing an appropriate level of conservativeness is very important for effective exploration in biological sequence design. A fixed value of the conservativeness parameter Œ¥ may be not optimal, as proxy models can be confident in some regions of the sequence space while highly uncertain in others. To address this, the proposed method introduces adaptive conservativeness, where the extent of exploration is dynamically adjusted based on the proxy model‚Äôs uncertainty. The conservativeness parameter is defined as a function of each sequence.

$$
\delta(x;\sigma) = \delta_{\mathrm{const.}} - \lambda \sigma(x)
$$

where œÉ(x) denotes the uncertainty of the proxy model's prediction for sequence x, Œ¥<sub>const</sub> is a base conservativeness level, and Œª controls the strength of the adjustment. The resulting value is clamped to remain within [0,1]. This encourages more conservative search when the proxy is uncertain, restricting exploration to nearby, reliable regions, while allowing broader exploration when the proxy is confident. By aligning the search behaviour with model confidence, adaptive conservativeness improves effectiveness against proxy misspecification and leads to more stable and effective policy training.


## Experimental Setup and Benchmarks

To validate the effectiveness of our Œ¥-Conservative Search (Œ¥-CS), we conducted extensive experiments across a variety of biological sequence design tasks, including DNA, RNA, protein, and peptide design. Our goal was to evaluate whether Œ¥-CS could consistently improve upon existing methods, particularly in settings where proxy models are prone to failure. Firstly, we investigate proxy failures and the impact of our conservative search by applying Œ¥-CS to GFN-AL. Secondly, we evaluate our algorithm's performace and compare it to other baselines.

### 1. Hard TF-Bind 8

The task is called Hard TF-Bind 8. In this task we aim to generate DNA sequences (length L= 8) having maximize binding affinity. We collect the initial dataset near a certain sequence. The sequences in initial dataset have lower scores than the given sequence. We then form a dataset D0 of size 1,024. Also we modify the score landscape to assign zero to sequence scoring < 0.3. It follows Real world design where search space is vast with limited data availability.

### 2. FLEXS Benchmark

In this experiment we compare the performace of our algorithm with other baselines. The architecture of the proxy model used is as follows. We employ a convolutional neural network (CNN) with one-dimensional convolutions according to <a href="#ref-4" title="(2020) AdaLead: A simple and robust adaptive greedy search algorithm for sequence design">[4]</a>. We use a UCB acquisition function and measure the uncertainty with an ensemble of three network instances. We use representative exploration algorithms and use same architecture to implement proxy models for all baselines, except for GFlowNets Active Learning which uses original implementation.

<b>Baselines :-</b>

<b>AdaLead : </b> It is an evolutionary type method that employs adaptive greedy search with hill climbing <a href="#ref-4" title="(2020) AdaLead: A simple and robust adaptive greedy search algorithm for sequence design">[4]</a>.

<b>Bayesian optimization : </b> Black-box optimization algorithm.

<b>TuRBO : </b> It uses Gaussian process for local Bayesian optimization.

<b>CMA-ES : </b> It is an evolutionary type method that adapts covariance matrix evolution strategy.

<b>CbAS and DbAS : </b> It is a model based technique which has variational auto encoder with trust regions.

<b>DyNA PPO : </b> It is an On-policy RL method for proximal policy optimization with proxy <a href="#ref-8" title="(2020) Model-based reinforcement learning for biological sequence design. In International Conference on Learning Representations (ICLR)">[8]</a>.

<b>GFN-AL : </b> It is an Off-policy RL method which has GFlowNets with Bayesian active learning <a href="#ref-2" title="(2022) Biological sequence design with GFlowNets. In International Conference on Machine Learning (ICML)">[2]</a>.

For each task, we conduct 10 active learning rounds. Starting from the initial dataset D<sub>0</sub> with all of the query batch size as 128. To evaluate the performance, we measure the maximum, median, and mean scores of Top K sequences.

#### 2.1 RNA Sequence design

The goal is to design an RNA sequence that binds to the target with lowest binding energy. We measure the quality by ViennaRNA algorithm <a href="#ref-5" title="(2011) ViennaRNA package 2.0. Algorithms for molecular biology">[5]</a>. The length (L) of RNA is set to 14, with 4 tokens. There are three RNA binding tasks, namely RNA-A, RNA-B, and RNA-C.  Initial datasets consist of 5,000 randomly generated sequences and offline dataset provided in <a href="#ref-6" title="(2023) Bootstrapped training of score-conditioned generator for offline design of biological sequences. In Advances in Neural Information Processing Systems (NeurIPS)">[6]</a>. Œ¥ is set to 0.5 and Œª is set to 5.

#### 2.2 DNA Sequence Design

For this task we follow the experimental setup from previous setup <a href="#ref-4" title="(2020) AdaLead: A simple and robust adaptive greedy search algorithm for sequence design">[4]</a><a href="#ref-2" title="(2022) Biological sequence design with GFlowNets. In International Conference on Machine Learning (ICML)">[2]</a><a href="#ref-7" title="(2022) DesignBench: Benchmarks for data-driven offline model-based optimization. In International Conference on Machine Learning (ICML)">[7]</a><a href="#ref-6" title="(2023) Bootstrapped training of score-conditioned generator for offline design of biological sequences. In Advances in Neural Information Processing Systems (NeurIPS)">[6]</a>. The Initial dataset D<sub>0</sub> is taken as the bottom 50% in the scoring landscape. Which have nearly 33,000 samples and the maximum score is 0.4. Œ¥ is set to 0.5 and Œª is set to 5.

#### 2.3 Protein Sequence Design

We take two tasks for protien sequence design. Green fluorescent Protein (GFP) <a href="#ref-9" title="(2016) Local fitness landscape of the green fluorescent protein. Nature">[9]</a> and Additive adeno-associated virus (AAV)<a href="#ref-10" title="(2019) Comprehensive AAV capsid fitness landscape reveals a viral gene and enables machine-guided design. Science">[10]</a>. GFP is a protein that has green fluorescence when it is exposed to light in ultraviolet range, and the objective of this task is to identify protein sequences with high log-fluorescence intensity values. AAV are small virus that affects humans and some other species. Many features make AAV an attractive candidate for gene therapy. In this task our aim is to discover sequences that lead to higher gene therapeutic efficiency.

The vocabulary for both the tasks are defined as standard set of 20 amino acids, the initial dataset D<sub>0</sub> is generated by random mutation , we set Œ¥ to 0.05 and Œª is set to 1. For GFP task, the sequence length (L) is 238 while for AAV task it is 90. The initial dataset for GFP task has approx. 10,200 sequences with a maximum score of 3.5 while for AAV task it is approx. 15,300 with maximum score being 0.5. 

## Results

### Experiment 1 : Hard TF-Bind 8

We noticed that the proxy provides very good results, it gives accurate predictions for data points in the initial dataset D<sub>0</sub>, the red points shows the results for initial dataset in Fig.2. However, if me move further from our observed initial dataset, the yielded results are not good. It is producing unreliable predictions for points outside D<sub>0</sub> or for out-of-distribution data (OOD). The correlation between the true oracle and proxy model estimation function decreases as we move farther in search space. So, here the hypothesis holds true, that "the proxy models perform poorly for out-of-distribution (OOD) Data".

![Figure 2: Proxy failure on Hard TF-Bind-8.]({{ site.baseurl }}/images/Fig2.png)
*Figure 2: Proxy failure on Hard TF-Bind-8.*

When we introduced Œ¥-CS in this experiment, we clearly see the results improving (Fig.3). If we adjust Œ¥<1, the GFlowNets Active Learning performs better and yields high quality sequences with greater correlation with true score values. It is restricting the serach space to smaller regions near the high quality sequences.

![Figure 3: Mean score over rounds on Hard TF-Bind-8.]({{ site.baseurl }}/images/Fig3.png)
*Figure 3: Mean score over rounds on Hard TF-Bind-8.*

### Experiment 2 : FLEXS Benchmark

**RNA Sequence Design**

Results shows that our algorithm is outperforming all other baselines for RNA designing tasks. We have three types RNA-A, RNA-B and RNA-C, in all the three types our method with GFlowNets Active Learning with Œ¥-CS is standing as the top performer. We can see the comparison in Fig.4 for RNA-A, Fig.5 for RNA-B and Fig.6 for RNA-C. The values depicted in the graph are available in the tables below the graph respectively.

![Figure 4: The max, median, and mean curve over rounds in RNA-A.]({{ site.baseurl }}/images/Fig4.png)
*Figure 4: The max, median, and mean curve over rounds in RNA-A.*

*Table 1. Results on RNA-A after ten rounds.*

| Method          | Max               | Median            | Mean              | Diversity         | Novelty          |
|-----------------|-------------------|-------------------|-------------------|-------------------|------------------|
| AdaLead         | 0.968 ¬± 0.070     | 0.808 ¬± 0.049     | 0.817 ¬± 0.048     | 3.518 ¬± 0.446     | 6.888 ¬± 0.426    |
| BO              | 0.722 ¬± 0.025     | 0.510 ¬± 0.008     | 0.528 ¬± 0.004     | **9.531 ¬± 0.062** | 5.842 ¬± 0.083    |
| TuRBO           | 0.875 ¬± 0.078     | 0.670 ¬± 0.093     | 0.682 ¬± 0.096     | 3.695 ¬± 0.166     | 6.464 ¬± 0.759    |
| CMA-ES          | 0.816 ¬± 0.030     | 0.585 ¬± 0.016     | 0.599 ¬± 0.020     | 5.747 ¬± 0.110     | 6.373 ¬± 0.159    |
| CbAS            | 0.678 ¬± 0.020     | 0.467 ¬± 0.009     | 0.481 ¬± 0.008     | 9.457 ¬± 0.189     | 5.428 ¬± 0.078    |
| DbAS            | 0.670 ¬± 0.041     | 0.472 ¬± 0.016     | 0.485 ¬± 0.015     | 9.483 ¬± 0.100     | 5.450 ¬± 0.132    |
| DyNA PPO        | 0.737 ¬± 0.022     | 0.507 ¬± 0.007     | 0.521 ¬± 0.009     | 8.889 ¬± 0.034     | 5.828 ¬± 0.095    |
| GFN-AL          | 1.030 ¬± 0.024     | 0.838 ¬± 0.013     | 0.849 ¬± 0.013     | 6.983 ¬± 0.159     | 7.398 ¬± 0.024    |
| GFN-AL + Œ¥-CS   | **1.055 ¬± 0.000** | **0.939 ¬± 0.008** | **0.947 ¬± 0.009** | 6.442 ¬± 0.525     | **7.406 ¬± 0.066**|



![Figure 5: The max, median, and mean curve over rounds in RNA-B.]({{ site.baseurl }}/images/Fig5.png)
*Figure 5: The max, median, and mean curve over rounds in RNA-B.*

*Table 2. Results on RNA-B after ten rounds.*

| Method          | Max               | Median            | Mean              | Diversity         | Novelty          |
|-----------------|-------------------|-------------------|-------------------|-------------------|------------------|
| AdaLead         | 0.965 ¬± 0.033     | 0.817 ¬± 0.036     | 0.828 ¬± 0.032     | 3.334 ¬± 0.423     | 7.441 ¬± 0.135    |
| BO              | 0.720 ¬± 0.032     | 0.502 ¬± 0.013     | 0.517 ¬± 0.014     | **9.495 ¬± 0.103** | 5.903 ¬± 0.116    |
| TuRBO           | 0.869 ¬± 0.086     | 0.705 ¬± 0.073     | 0.715 ¬± 0.073     | 3.638 ¬± 0.173     | 6.713 ¬± 0.560    |
| CMA-ES          | 0.850 ¬± 0.063     | 0.581 ¬± 0.028     | 0.602 ¬± 0.032     | 5.568 ¬± 0.365     | 6.480 ¬± 0.200    |
| CbAS            | 0.668 ¬± 0.021     | 0.465 ¬± 0.005     | 0.477 ¬± 0.004     | 9.234 ¬± 0.356     | 5.523 ¬± 0.083    |
| DbAS            | 0.652 ¬± 0.021     | 0.463 ¬± 0.019     | 0.475 ¬± 0.019     | 9.019 ¬± 0.648     | 5.537 ¬± 0.150    |
| DyNA PPO        | 0.730 ¬± 0.088     | 0.481 ¬± 0.028     | 0.499 ¬± 0.029     | 8.978 ¬± 0.196     | 5.839 ¬± 0.198    |
| GFN-AL          | 1.001 ¬± 0.016     | 0.858 ¬± 0.004     | 0.870 ¬± 0.006     | 6.599 ¬± 0.384     | **7.673 ¬± 0.043**|
| GFN-AL + Œ¥-CS   | **1.014 ¬± 0.001** | **0.929 ¬± 0.004** | **0.934 ¬± 0.003** | 5.644 ¬± 0.307     | 7.661 ¬± 0.064    |


![Figure 6: The max, median, and mean curve over rounds in RNA-C.]({{ site.baseurl }}/images/Fig6.png)
*Figure 6: The max, median, and mean curve over rounds in RNA-C.*

*Table 3. Results on RNA-C after ten rounds.*

| Method          | Max               | Median            | Mean              | Diversity         | Novelty          |
|-----------------|-------------------|-------------------|-------------------|-------------------|------------------|
| AdaLead         | 0.867 ¬± 0.081     | 0.723 ¬± 0.057     | 0.735 ¬± 0.057     | 3.893 ¬± 0.444     | 5.856 ¬± 0.515    |
| BO              | 0.694 ¬± 0.034     | 0.506 ¬± 0.003     | 0.519 ¬± 0.003     | 9.714 ¬± 0.054     | 5.430 ¬± 0.043    |
| TuRBO           | 0.773 ¬± 0.222     | 0.621 ¬± 0.221     | 0.632 ¬± 0.215     | 4.403 ¬± 1.542     | 4.906 ¬± 1.701    |
| CMA-ES          | 0.753 ¬± 0.062     | 0.496 ¬± 0.041     | 0.521 ¬± 0.037     | 5.581 ¬± 0.399     | 5.019 ¬± 0.294    |
| CbAS            | 0.696 ¬± 0.041     | 0.492 ¬± 0.018     | 0.507 ¬± 0.017     | **9.518 ¬± 0.310** | 5.033 ¬± 0.086    |
| DbAS            | 0.678 ¬± 0.025     | 0.495 ¬± 0.010     | 0.508 ¬± 0.011     | 9.249 ¬± 0.414     | 5.128 ¬± 0.153    |
| DyNA PPO        | 0.728 ¬± 0.060     | 0.478 ¬± 0.015     | 0.489 ¬± 0.015     | 9.246 ¬± 0.086     | 5.306 ¬± 0.124    |
| GNF-AL          | 0.951 ¬± 0.034     | 0.774 ¬± 0.004     | 0.786 ¬± 0.004     | 7.072 ¬± 0.163     | **6.661 ¬± 0.071**|
| GFN-AL + Œ¥-CS   | **1.094 ¬± 0.045** | **0.972 ¬± 0.043** | **0.983 ¬± 0.043** | 6.493 ¬± 1.751     | 6.494 ¬± 0.084    |


**DNA Sequence Design**
In this task, we see in the results that GFN-AL is performing well without Œ¥-CS. However, when we intergrate Œ¥-CS with GFN-AL, it performs slighly better that all other baselines (Fig.7).

![Figure 7: The max, median, and mean curve over rounds in DNA (TF-Bind-8).]({{ site.baseurl }}/images/Fig7.png)
*Figure 7: The max, median, and mean curve over rounds in DNA (TF-Bind-8).*

*Table 4. Results on DNA (TF-Bind 8) after ten rounds.*

| Method          | Max               | Median            | Mean              | Diversity         | Novelty          |
|-----------------|-------------------|-------------------|-------------------|-------------------|------------------|
| AdaLead         | **0.995 ¬± 0.004** | 0.937 ¬± 0.008     | 0.939 ¬± 0.007     | 3.506 ¬± 0.267     | 1.194 ¬± 0.035    |
| BO              | 0.977 ¬± 0.008     | 0.806 ¬± 0.007     | 0.815 ¬± 0.005     | 4.824 ¬± 0.074     | 1.144 ¬± 0.029    |
| TuRBO           | 0.944 ¬± 0.039     | 0.719 ¬± 0.091     | 0.725 ¬± 0.097     | 4.276 ¬± 0.680     | 1.005 ¬± 0.182    |
| CMA-ES          | 0.986 ¬± 0.008     | 0.843 ¬± 0.032     | 0.843 ¬± 0.030     | 3.617 ¬± 0.321     | 1.130 ¬± 0.083    |
| CbAS            | 0.988 ¬± 0.004     | 0.835 ¬± 0.011     | 0.845 ¬± 0.009     | 4.662 ¬± 0.079     | 1.134 ¬± 0.021    |
| DbAS            | 0.987 ¬± 0.004     | 0.831 ¬± 0.005     | 0.845 ¬± 0.005     | 4.694 ¬± 0.056     | 1.141 ¬± 0.047    |
| DyNA PPO        | 0.977 ¬± 0.013     | 0.746 ¬± 0.010     | 0.761 ¬± 0.006     | 4.430 ¬± 0.030     | 1.120 ¬± 0.021    |
| GFN-AL          | 0.976 ¬± 0.002     | 0.947 ¬± 0.004     | 0.947 ¬± 0.009     | 3.158 ¬± 0.166     | **2.409 ¬± 0.071**|
| GFN-AL + Œ¥-CS   | 0.981 ¬± 0.002     | **0.971 ¬± 0.006** | **0.972 ¬± 0.005** | **1.277 ¬± 0.182** | 2.237 ¬± 0.356    |


**Protien Sequence Design**

In this task, the search space was very large. The results shows that most of the baselines perform poorly in identifying new sequences that are better in quality than the highest quality sequence in the dataset. But, our method find high scoring sequences outside of dataset from round 1 (Fig.8 and Fig.9).

![Figure 8: The max, median, and mean curve over rounds in GFP.]({{ site.baseurl }}/images/Fig8.png)
*Figure 8: The max, median, and mean curve over rounds in GFP.*

*Table 5. Results on GFP after ten rounds.*

| Method          | Max               | Median            | Mean              | Diversity         | Novelty           |
|-----------------|-------------------|-------------------|-------------------|-------------------|-------------------|
| AdaLead         | 3.581 ¬± 0.004     | 3.549 ¬± 0.002     | 3.552 ¬± 0.002     | 47.237 ¬± 1.213    | 1.467 ¬± 0.094     |
| BO              | 3.572 ¬± 0.000     | 3.378 ¬± 0.000     | 3.331 ¬± 0.000     | **62.955 ¬± 0.000**| 0.000 ¬± 0.000     |
| TuRBO           | 3.581 ¬± 0.006     | 3.466 ¬± 0.049     | 3.438 ¬± 0.057     | 51.112 ¬± 6.802    | 0.584 ¬± 0.379     |
| CMA-ES          | 3.572 ¬± 0.000     | 3.410 ¬± 0.000     | 3.384 ¬± 0.000     | 58.299 ¬± 0.000    | 0.000 ¬± 0.010     |
| CbAS            | 3.572 ¬± 0.000     | 3.378 ¬± 0.100     | 3.334 ¬± 0.002     | 62.926 ¬± 0.139    | 0.009 ¬± 0.012     |
| DbAS            | 3.572 ¬± 0.000     | 3.378 ¬± 0.200     | 3.334 ¬± 0.002     | 62.926 ¬± 0.139    | 0.009 ¬± 0.012     |
| DyNA PPO        | 3.572 ¬± 0.000     | 3.378 ¬± 0.300     | 3.331 ¬± 0.000     | **62.955 ¬± 0.100**| 0.000 ¬± 0.010     |
| GFN-AL          | 3.578 ¬± 0.003     | 3.511 ¬± 0.006     | 3.508 ¬± 0.004     | 60.278 ¬± 0.819    | **20.837 ¬± 0.916**|
| GFN-AL + Œ¥-CS   | **3.592 ¬± 0.003** | **3.567 ¬± 0.003** | **3.569 ¬± 0.003** | 46.255 ¬± 10.534   | 17.459 ¬± 5.538    |



![Figure 9: The max, median, and mean curve over rounds in AAV.]({{ site.baseurl }}/images/Fig9.png)
*Figure 9: The max, median, and mean curve over rounds in AAV.*

*Table 6. Results on AAV after ten rounds.*

| Method          | Max               | Median            | Mean              | Diversity          | Novelty           |
|-----------------|-------------------|-------------------|-------------------|--------------------|-------------------|
| AdaLead         | 0.565 ¬± 0.027     | 0.505 ¬± 0.016     | 0.509 ¬± 0.017     | 5.693 ¬± 0.946      | 2.133 ¬± 1.266     |
| BO              | 0.500 ¬± 0.000     | 0.478 ¬± 0.000     | 0.480 ¬± 0.000     | 4.536 ¬± 0.000      | 0.000 ¬± 0.000     |
| TuRBO           | 0.500 ¬± 0.000     | 0.548 ¬± 0.000     | 0.540 ¬± 0.000     | 4.533 ¬± 0.006      | 0.002 ¬± 0.003     |
| CMA-ES          | 0.500 ¬± 0.000     | 0.581 ¬± 0.000     | 0.582 ¬± 0.000     | 4.148 ¬± 0.000      | 0.000 ¬± 0.010     |
| CbAS            | 0.500 ¬± 0.000     | 0.598 ¬± 0.000     | 0.600 ¬± 0.000     | 4.545 ¬± 0.018      | 0.002 ¬± 0.003     |
| DbAS            | 0.500 ¬± 0.000     | 0.610 ¬± 0.000     | 0.620 ¬± 0.000     | 4.545 ¬± 0.018      | 0.002 ¬± 0.003     |
| DyNA PPO        | 0.500 ¬± 0.000     | 0.630 ¬± 0.000     | 0.640 ¬± 0.000     | 4.536 ¬± 0.010      | 0.000 ¬± 0.000     |
| GFN-AL          | 0.560 ¬± 0.008     | 0.609 ¬± 0.002     | 0.613 ¬± 0.002     | 4.044 ¬± 0.303      | 1.966 ¬± 0.157     |
| GFN-AL + Œ¥-CS   | **0.708 ¬± 0.010** | **0.663 ¬± 0.007** | **0.665 ¬± 0.006** | **11.296 ¬± 0.865** | **10.233 ¬± 0.822**|


## Conclusion



<a id="ref-1"></a>
<a id="ref-2"></a>
<a id="ref-3"></a>
<a id="ref-4"></a>
<a id="ref-5"></a>
<a id="ref-6"></a>
<a id="ref-7"></a>
<a id="ref-8"></a>
<a id="ref-9"></a>
<a id="ref-10"></a>
<a id="ref-11"></a>
<a id="ref-12"></a>
<a id="ref-13"></a>
<a id="ref-14"></a>

## References

1. Hyeonah Kim, Minsu Kim, Taeyoung Yun, Sanghyeok Choi, Emmanuel Bengio, Alex Hern√°ndez-Garc√≠a, Jinkyoo Park (2025). Improved Off-policy Reinforcement Learning in Biological Sequence Design. [https://arxiv.org/abs/2102.09844  ](https://arxiv.org/abs/2410.04461)
2. Jain, M., Bengio, E., Hernandez-Garcia, A., Rector-Brooks, J., Dossou, B. F., Ekbote, C. A., Fu, J., Zhang, T., Kilgour, M., Zhang, D., et al. (2022) Biological sequence design with GFlowNets. In International Conference on Machine Learning (ICML).  
3. Tripp, A., Daxberger, E., and Hern¬¥ andez-Lobato, J. M. (2020) Sample-efficient optimization in the latent space of deep generative models via weighted retraining. In Advances in Neural Information Processing Systems (NeurIPS).  
4. Sinai, S., Wang, R., Whatley, A., Slocum, S., Locane, E.,and Kelsic, E. D. (2020) AdaLead: A simple and robust adaptive greedy search algorithm for sequence design. 
5. Lorenz, R., Bernhart, S. H., H¬® oner zu Siederdissen, C., Tafer, H., Flamm, C., Stadler, P. F., and Hofacker, I. L. (2011) ViennaRNA package 2.0. Algorithms for molecular biology,6:1‚Äì14.
6. Kim, M., Berto, F., Ahn, S., and Park, J. (2023) Bootstrapped training of score-conditioned generator for offline design of biological sequences. In Advances in Neural Information Processing Systems (NeurIPS).  
7. Trabucco, B., Geng, X., Kumar, A., and Levine, S. (2022) DesignBench: Benchmarks for data-driven offline model-based optimization. In International Conference on Machine Learning (ICML).
8. Angermueller, C., Dohan, D., Belanger, D., Deshpande, R., Murphy, K., and Colwell, L. (2020) Model-based reinforcement learning for biological sequence design. In International Conference on Learning Representations (ICLR).
9. Sarkisyan, K. S., Bolotin, D. A., Meer, M. V., Usmanova, D. R., Mishin, A. S., Sharonov, G. V., Ivankov, D. N.,Bozhanova, N. G., Baranov, M. S., Soylemez, O., et al. (2016) Local fitness landscape of the green fluorescent protein. Nature, 533(7603):397‚Äì401.
10. Ogden, P. J., Kelsic, E. D., Sinai, S., and Church, G. M. (2019) Comprehensive AAV capsid fitness landscape reveals a viral gene and enables machine-guided design. Science, 366(6469):1139‚Äì1143.


